Making the instrumented clinit_wrapper function thread-safe

Degiorgio opened the pull request:
The previous version of clinit_wrapper did not conform to the JVM specification (section 5.5) as it did not support concurrency. More specifically, the clinit_wrapper has to be carefully synchronized
as other threads may try to initialize a given class or interface at the same-time.

This commit **adds** a new implementation  of the clinit_wrapper  that is thread-safe. This is achieved by introducing a critical section and two state variables. The commit implements a simplification of the algorithm defined in section 5.5 of the JVM spec. For instance exceptions thrown during the execution of static initializers are not properly handled.  To use this implementation instead of the old one, a new option has been added 'java-threading'.  We made this opt-in  instead of simply replacing  the old one for performance reasons.

Commit  duplicates some  regression tests, while these tests do not exhibit concurrency, the resulting change to the clinit_wrapper trigged by specifying 'java-threading' effects non-concurrent code.

Note: this change is required in-order to add concurrency support to the java-front-end.

tautschnig committed:
fixup! Support __attribute__((section("x"))

marek-trtik committed:
Fixing member offset computation in presence of bitfields

The previous algorithm correctly identified the offset of the next non-bitfield
member, but would fail on consecutive bit-fields (fitting in one byte) as the
offset was eagerly incremented. Also cleanup this computation in all places that
used a similar algorithm.

thk123 commented:
. @smowton should definitely look at this as has been touching this recently, I'd also appreciate a test-gen pointer bump.

peterschrammel commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -972,109 +972,277 @@ bool java_bytecode_convert_methodt::class_needs_clinit(
+/// This wrapper is called immediately before any read or write of a static
+/// field or whenever an instance of the class is created.
+/// When class \p classname does not need static initialization we
+/// return a nil irep.  We also explicitely avoid introducing calls to the
```
explicitly

peterschrammel commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -972,109 +972,277 @@ bool java_bytecode_convert_methodt::class_needs_clinit(
+
+  // Initialize the syper-class C' and
+  // the implemented interfaces l_1 ... l_n.
+  // see JVM Spec p.359 step 7, for the exact defination of
```
definition

peterschrammel commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -3112,3 +3283,68 @@ void java_bytecode_convert_methodt::create_stack_tmp_var(
+  const clinit_statest state)
+{
+  mp_integer initv(static_cast<int>(state));
+  constant_exprt init_s=from_integer(initv, java_int_type());
```
Do you require a full `int` here?

peterschrammel commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -972,109 +972,277 @@ bool java_bytecode_convert_methodt::class_needs_clinit(
-/// place, calls the actual <clinit> method if not, and initializes super-
-/// classes and interfaces.
-/// \param classname: Class name
+/// This function retrieves from or creates into the symbol table a wrapper
```
It might be worth factoring out the clinit generation into a separate .cpp file in order to reduce the size of this monster.cpp.

Degiorgio replied:
agreed, will do this.

smowton replied:
Hang on! This is already done in the PR this is racing against :)

smowton commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -972,109 +972,277 @@ bool java_bytecode_convert_methodt::class_needs_clinit(
+  // if we are initializing class C due to a use of class C within method
+  // `C.<clinit>` then avoid the recursive (and unnecessary) call to
+  // the wrapper.
+  const irep_idt &real_clinit_name=id2string(classname)+".<clinit>:()V";
```
Since this is new code, use the new clang-format-compatible / human-readable operator spacing (i.e. spaces around ==, =, etc)

smowton commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -972,109 +972,277 @@ bool java_bytecode_convert_methodt::class_needs_clinit(
+    true, false);
+
+  code_blockt function_body;
+  codet atomic_begin(ID_atomic_begin);
```
Consider making a `std_code` entry for these?

smowton commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -972,109 +972,277 @@ bool java_bytecode_convert_methodt::class_needs_clinit(
+    function_body.add(conditional);
+  }
+
+  // Initialize the syper-class C' and
```
superclass

smowton commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -972,109 +972,277 @@ bool java_bytecode_convert_methodt::class_needs_clinit(
+  // This is achieved  by iterating through the base types and
+  // adding recursive calls to the clinit_wrapper()
+  {
+    const symbolt &class_symbol=*symbol_table.lookup(classname);
```
lookup_ref

smowton commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -3112,3 +3283,68 @@ void java_bytecode_convert_methodt::create_stack_tmp_var(
+  const irep_idt &name,
+  const typet &type,
+  const exprt &value,
+  const bool is_thread_local,
```
Generally base_name is the unqualified name, e.g. `name` == `java::Class.field` => `base_name` == `field`.

smowton replied:
These aren't very obvious at the callsite -- perhaps use two-valued enums so we don't mix up `true, false` and `false, true`?

smowton commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -3112,3 +3283,68 @@ void java_bytecode_convert_methodt::create_stack_tmp_var(
+/// /param is_thread_local: if true this symbol will be set as thread local
+/// /param is_static_lifetime: if true this symbol will be set as static
+/// /return returns new symbol.
+symbolt java_bytecode_convert_methodt::add_new_symbol(
```
This thing is rather specific! Probably rename it "add_variable_symbol"?

smowton commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -3112,3 +3283,68 @@ void java_bytecode_convert_methodt::create_stack_tmp_var(
+///   equal exprt.
+/// /return returns a equal_exprt, equating \p symbol_expr to the integer
+///   representation of \p state
+equal_exprt java_bytecode_convert_methodt::gen_clinit_eqexpr(
```
These don't really need to be members -- they're clearly `static` helpers

smowton reviewed (commented):
Some nitpicks, and some bad news: this is going to clash with https://github.com/diffblue/cbmc/pull/1774, which I hope to get in later today, and reworks how static init methods are generated to create them lazily, and create their state variables (the bool `already_run`, and now your new `state` variable) before method conversion.

The good news however: this should be quite easy to rebase on #1774 -- simply create your additional global where #1774 does and copy-paste your init code into my new `java_static_initializers.cpp` file.

The **local** variable should be created during method elaboration (i.e. in `get_clinit_wrapper_body`).

Degiorgio commented:
@smowton, oks will rebase on top of 1774

romainbrenguier committed:
Use string-max-length as default max input-length

Having input string longer than string-max-length does not make sense as
the solver will not know how to analyse them.
So when string-max-input-length is not specified we can use
string-max-length instead.

Degiorgio committed:
Modifies the instrumented clinit_wrapper function

The previous version of clinit_wrapper did not conform to the JVM
specification (section 5.5) as it did not support concurrency. More
specifically, the clinit_wrapper has to be carefully synchronised
as other threads may try to initialize a given class or interface
at the same-time.

This commit makes the clinit_wrapper thread-safe by introducing a
critical section and two state variables. The commit implements a
simplification of the algorithm defined in section 5.5 of the JVM
specification. For instance, exceptions thrown during the execution
of static initializers are not properly handled.

Degiorgio commented:
@smowton , rebase done, can you take a look please?

smowton commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -8,11 +8,38 @@ Author: Chris Smowton, chris.smowton@diffblue.com
+{
+  NOT_INIT,
+  IN_PROGRESS,
+  RETURN
```
INIT_COMPLETE?

Degiorgio replied:
instead of 'RETURN'? yes INIT_COMPLETE is better, changing.

cesaro replied:
@Degiorgio Do not forget to update the documentation above

smowton commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -134,82 +260,239 @@ static void create_clinit_wrapper_symbols(
-  equal_exprt check_already_run(
-    already_run_symbol.symbol_expr(),
-    false_exprt());
+  const symbolt &clinit_s_sym =
```
Use `_state` instead of `_s`

Degiorgio replied:
done

smowton commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -95,21 +202,40 @@ static void create_clinit_wrapper_symbols(
+
+  add_new_symbol(
+    symbol_table,
+    clinit_local_state_variable_name(class_name),
```
Rename this variable `_thread_local` rather than just `_local` to clarify we mean thread-locality not that it's a function local

Degiorgio replied:
done

smowton commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -134,82 +260,239 @@ static void create_clinit_wrapper_symbols(
+///     java::C::__CPROVER_clinit_state = IN_PROGRESS
+///     was_ready = false;
+///   }
+///   else If(java::C::__CPROVER_clinit_state == READY)
```
Out of date? AFAIK there is no state named READY anymore

Degiorgio replied:
yep this should be RETURN

smowton commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -134,82 +260,239 @@ static void create_clinit_wrapper_symbols(
+///   java::C::__CPROVER_clinit_local_state = RETURN;
+///
+///   ATOMIC_BEGIN
+///   Assume(java::C::__CPROVER_clinit_state != IN_PROGRESS)
```
?? Will that work? Presumably we want the other thread to busy-wait for IN_PROGRESS to transition to DONE in this case. If this is indeed magic that symex understands like a wait then write that in the comments.

cesaro replied:
Yes, that will work. It's exactly how POSIX mutexes are implemented in `pthread_mutex_lock`, see here:

https://github.com/diffblue/cbmc/blob/develop/src/ansi-c/library/pthread_lib.c#L132

Thanks, we will indeed document it.

smowton reviewed (commented):
Some suggestions

cesaro commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -95,21 +202,40 @@ static void create_clinit_wrapper_symbols(
+    true);
+
+  // Create a local variable "was_ready". This variable is used optimization as
+  // an optimization to make even more clear to the symex engine that it does
```
Typo: "is used as an optimization"

Degiorgio replied:
done

cesaro commented:
@smowton all suggestions applied, please have a another look

smowton commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -57,6 +107,64 @@ static irep_idt clinit_function_name(const irep_idt &class_name)
+///   equal exprt.
+/// /return returns a equal_exprt, equating \p symbol_expr to the integer
+///   representation of \p state
+equal_exprt
```
static?

Degiorgio replied:
done

Degiorgio commented:
@peterschrammel  can you take another  look at this please? All the issues you raised should be addressed.

smowton commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -96,21 +204,41 @@ static void create_clinit_wrapper_symbols(
+    symbol_table,
+    clinit_local_init_complete_var_name(class_name),
+    bool_typet(),
+    false_exprt(),
```
fwiw the `value` member of a local variable is ignored; it's up to you to generate an initialiser. Therefore recommend passing `nil_exprt()` there and checking in add-new-symbol that function-scoped (non-global) variables do not attempt to specify an initialiser this way.

Degiorgio replied:
done

smowton commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -135,82 +263,256 @@ static void create_clinit_wrapper_symbols(
+  static_assert((int) clinit_statest::INIT_COMPLETE==2, "Check comment above");
+#endif
+
+  /// bool init_complete;
```
Non-doxy (//) comment

Degiorgio replied:
done

smowton commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -96,21 +204,41 @@ static void create_clinit_wrapper_symbols(
+    true,
+    true);
+
+  // Create a local variable "init_complete". This variable is used as an
```
I think this comment is inaccurate, does it belong to the thread-local variable instead? The function-local one is just used to copy the global so you can inspect it and then maybe return outside a critical section.

Degiorgio replied:
the comment is misleading agreed. rewording.

Degiorgio replied:
We are using a local boolean for inspection on purpose. The reason being that inspecting the global state  resulted in symex related propagation issues in multi-threaded code.

smowton commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -96,21 +204,41 @@ static void create_clinit_wrapper_symbols(
+  // Create a local variable "init_complete". This variable is used as an
+  // optimization to make even more clear to the symex engine that it does not
+  // need to unwind ifinitely many copes of the wrapper
+  add_new_symbol(
```
Also the *local* should be created in the populate-function-body phase. Only globals should be created at this stage (i.e. the global, the thread-scoped variable, and the function symbol itself).

Degiorgio replied:
done

smowton reviewed (commented):
Sorry, more stuff to fix

Degiorgio commented:
@smowton latest commit should address the issues you raised.

tautschnig commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -135,82 +251,263 @@ static void create_clinit_wrapper_symbols(
+///   // ...
+///   java::In::clinit_wrapper();
+///
+///   java::C::<clinit>();
```
Doxygen isn't happy with the above:
```
/home/travis/build/diffblue/cbmc/src/java_bytecode/java_static_initializers.cpp:294: warning: explicit link request to 'clinit_wrapper()' could not be resolved
/home/travis/build/diffblue/cbmc/src/java_bytecode/java_static_initializers.cpp:300: warning: Unsupported xml/html tag <clinit> found
```

Degiorgio replied:
doxygen should be happy now

tautschnig committed:
Only the top-level section should be considered for renaming

The Linux kernel first places `jiffies64` in .data, and later refines
that to .data..cacheline_aligned. The latter still is within the .data
section, though, so just use the .data part for renaming.

cesaro commented:
A TG pointer bump is necessary.

thk123 commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -14,6 +14,34 @@ Author: Chris Smowton, chris.smowton@diffblue.com
+/// `INIT_COMPLETE` corresponds to "the `<clinit>` has been executed and the
+///   class is ready to be used, or it has errored"
+///
+/// The last state corresponds to a fusion of the two original states "ready
```
I'm sure there are good reasons - but why fuse them? Normally easier to understand code when it cleanly maps on to the spec. In any case, a comment explaining why would be good here.

Degiorgio replied:
For simplification reasons,  current implementation (also the previous one) does not handle exceptions (this stated in the comment for get_clinit_wrapper_body). so the 'errored" state is not really possible in current implementation.

thk123 commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -39,13 +67,35 @@ bool is_clinit_wrapper_function(const irep_idt &function_id)
+/// /param is_thread_local: if true this symbol will be set as thread local
+/// /param is_static_lifetime: if true this symbol will be set as static
+/// /return returns new symbol.
+static symbolt add_new_symbol(
```
This feels like the wrong place for this method - it amounts to a `symbolt` constructor and an add - I think I'd prefer a constructor or static factory method added to `symbolt`

thk123 replied:
Or perhaps using an existing method like `get_fresh_aux_symbol`?

thk123 commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -57,6 +107,64 @@ static irep_idt clinit_function_name(const irep_idt &class_name)
+/// \return static initializer wrapper-state variable global name
+static irep_idt clinit_state_var_name(const irep_idt &class_name)
+{
+  return id2string(class_name) + ".__CPROVER_clinit_state";
```
Nit: `return id2string(class_name) + CPROVER_PREFIX "state"`

thk123 commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -57,6 +107,64 @@ static irep_idt clinit_function_name(const irep_idt &class_name)
+/// \return static initializer wrapper-state local state variable name
+static irep_idt clinit_thread_local_state_var_name(const irep_idt &class_name)
+{
+  return id2string(class_name) + ".__CPROVER_clinit_threadlocal_state";
```
Ditto about using `CPROVER_PREFIX`

thk123 commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -57,6 +107,64 @@ static irep_idt clinit_function_name(const irep_idt &class_name)
+/// \return static initializer wrapper-state local variable
+static irep_idt clinit_local_init_complete_var_name(const irep_idt &class_name)
+{
+  return id2string(class_name) + ".__CPROVER_clinit_wrapper.init_complete";
```
Ditto about using `CPROVER_PREFIX`

thk123 commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -57,6 +107,64 @@ static irep_idt clinit_function_name(const irep_idt &class_name)
+///   representation of \p state
+static code_assignt
+gen_clinit_assignexpr(const exprt &symbol_expr, const clinit_statest state)
+{
```
Consider a precondition that `symbol_expr.type() == clinit_states_type`

thk123 replied:
Also use `symbol_exprt` for the first param (or don't call it `symbol_expr` for the name if that isn't actually required.

thk123 replied:
This is still outstanding - I think calling a variable `symbol_expr` but not requiring it to be a `symbol_exprt` is a confusing state - I see no reason for it not to be called simply `lhs`

thk123 commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -57,6 +107,64 @@ static irep_idt clinit_function_name(const irep_idt &class_name)
+/// /return returns a equal_exprt, equating \p symbol_expr to the integer
+///   representation of \p state
+static equal_exprt
+gen_clinit_eqexpr(const exprt &symbol_expr, const clinit_statest state)
```
Ditto: probably rename the param to by `expr` rather than `symbol_expr` if you don't require it to be a `symbol_expr`

thk123 commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -96,21 +204,29 @@ static void create_clinit_wrapper_symbols(
-  bool failed = symbol_table.add(already_run_symbol);
-  INVARIANT(!failed, "clinit-already-run symbol should be fresh");
-
+  exprt zero = from_integer(
```
I'd call this `not_init_value`

thk123 commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -135,82 +251,266 @@ static void create_clinit_wrapper_symbols(
+///   {
+///     return;
+///   }
+///   java::C::__CPROVER_clinit_thread_local_state = INIT_COMPLETE;
```
Why do we set it to be complete before running it?

thk123 replied:
Oh I see this is the local state - not sure the relevance of this TBH but perhaps it will become clear

thk123 replied:
I remain unsure why a thread local variable needs to track whether the init is complete (particularly if it is set to complete before completing...)

Degiorgio replied:
This the local thread state, look at the conditional statement just above it, which is inspecting it.  keep in mind that we do not handle exceptions, so when a thread  takes control of the initialization process  there can only be one outcome for the local state, which is INIT_COMPLETE.

thk123 commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -135,82 +251,266 @@ static void create_clinit_wrapper_symbols(
+///
+///  \endcode
+///
+/// Note: The current implementation does not deal with exceptions.
```
This seems fairly serious - is there a ticket to fix this (if so, could you provide a reference here)

Degiorgio replied:
The previous version also didn't handle exceptions. There is a performance vs soundness  balancing act aspect here as you have to be careful not to blowup the SAT formula. Having said that I am all for creating an issue if you feel this needs to be investigated further.

thk123 replied:
As your in the area I will trust your judgement on this matter :+1:

thk123 commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -135,82 +251,266 @@ static void create_clinit_wrapper_symbols(
+    function_body.add(decl);
+  }
+
+  // if(C::__CPROVER_local_state == INIT_COMPLETE) return;
```
This style is superb :+1:

thk123 commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -135,82 +251,266 @@ static void create_clinit_wrapper_symbols(
-    call_base.function() = findit->second.symbol_expr();
-    init_body.move_to_operands(call_base);
+    code_ifthenelset conditional;
+    code_blockt thenblk;
```
`then_block`

Degiorgio replied:
done

thk123 commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -135,82 +251,266 @@ static void create_clinit_wrapper_symbols(
-    code_function_callt call_base;
-    call_base.function() = findit->second.symbol_expr();
-    init_body.move_to_operands(call_base);
+    code_ifthenelset conditional;
```
call this `not_init_conditional`

Degiorgio replied:
done

thk123 commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -135,82 +251,266 @@ static void create_clinit_wrapper_symbols(
+    thenblk.add(code_assignt(init_complete.symbol_expr(), false_exprt()));
+    conditional.then_case() = thenblk;
+
+    code_ifthenelset condional_b;
```
and this `init_conditional` to disambiguate these two blocks

Degiorgio replied:
done

thk123 commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -135,82 +251,266 @@ static void create_clinit_wrapper_symbols(
+    function_body.add(conditional);
+  }
+
+  // Initialize the syper-class C' and
```
`syper` -> `super`

Degiorgio replied:
fixed.

thk123 reviewed (commented):
Nothing blocking but a bunch of suggestions, however (unless I've missed it in which case can you link it) this needs a TG pointer bump

Degiorgio commented:
@thk123  addressed issues you raised,  test-gen pointer bump: https://github.com/diffblue/test-gen/pull/1510

thk123 reviewed (commented):
All looks fine but not the TG pointer bump is currently failing - could you investigate why (might be that a rebase of this and the bump will fix it as it seems fairly unrelated to your changes)

romainbrenguier committed:
Get rid of default axioms for strings

These default axioms where too strict and where applied even for strings
that may never be created in the actual program.
This could for instance lead to problematic contradiction for a Java
object which could be casted either to a string or an int:
the axiom added there would apply both on the string and on the integer.

romainbrenguier committed:
Get rid of string_max_length field

In constraint generator, this was used for adding default axioms but is
no longer used.

Degiorgio commented:
Update, waiting for test-gen to pass CI

thk123 commented:
@Degiorgio sorry to keep banging on about this, but I really don't think having a variable called `symbol_expr` with a type `exprt` is a good idea. Either the type should be `symbol_exprt` or (more likely), the variable should be called `lhs` or `assign_to_expr`.

Degiorgio commented:
@thk123 , yes will change, I need to re-base anyway to get CI on test-gen pointer bump to pass,  as test-gen was broken earlier today.

tautschnig commented:
Did any of the merges I did this morning cause havoc? My apologies if that's the case, I was assuming people would call for test-gen bumps where there was risk.

tautschnig committed:
Reset namespace after symbolic execution

The state's symbol table may go out of scope, leaving the namespace with a
dangling pointer. This is a follow-up to e8eec2c8a8.

romainbrenguier committed:
Add test for generics array and strings

This adds an int greater than string-max-length in a generic array.
This is to check that TG-2138 is fixed.
Before the issue was solved setting the array to 101 was in
contradiction with default axioms added by the string solver.

romainbrenguier committed:
Add tests showing the effect on string-max-length

These tests show how the result of the solver can depend on the
string-max-input-length and string-max-length parameters.

tautschnig committed:
C front-end: Section/ASM renaming also needs to be applied to aliases

The example taken from the Linux kernel shows that people mix weak
aliases with section attributes. Not updating the alias name results in
a namespace lookup failure.

gcc_attributes11 is a reduced example taken from SV-COMP, where the same problem
arises.

romainbrenguier committed:
Use boolbvt for getting counter examples

We use this counter examples in the string solver but the formula given
there don't use arrays so it is enough to use boolbvt.

romainbrenguier committed:
Remove unnecessary replace_expr

This is already done in add_lemma.

JohnDumbell committed:
Increase AssertionError arguments allowed from 2 to 3

There is a constructor for this class with 2 arguments plus implied this.

tautschnig committed:
Add support for CaDiCaL

CaDiCaL is Armin Biere's latest solver with a minimal IPASIR-compatible
interface. In some initial experiments it considerable outperforms Minisat.

Commit:
Introduce nested exception printing

peterschrammel committed:
Remove java_bytecode deps from Makefiles

Commit:
Convert flatten_byte_extract to use structured exceptions

Store error in string to avoid dangling pointers

JohnDumbell committed:
Add a regression test.

JohnDumbell committed:
Update desc file to add pass variables.

Commit:
Introduce throwing bv_conversion expection

Remove suprisous namespace from bv_conversion_exceptions

Commit:
Introduce throwing a guard conversion exception

Throw the unwrapped exception as a string

danpoe committed:
Small shared two way pointer

Commit:
Provide the original goto statement in the error

This required the symex target_equation having the namespace to resolve
names

romainbrenguier committed:
Assign 0 to string length for null Java String

Although we should not reach this code if rhs is null, the association
`pointer -> length` is added to the solver anyway, so we have to make sure
the length of the string_exprt is set to something reasonable.

romainbrenguier committed:
Weaken invariant for nil exprt as model of array

nil_exprt happens when the underlying solver has no formula talking
about the array

romainbrenguier committed:
Make get_array return nil for long strings

Instead of aborting the program when the string is too long, we consider
this happened because of some invalid object, and return nil as a model
of the array.
Access to this invalid object is prevented by other part of the code and
should not occur in the trace.

romainbrenguier committed:
Correct bound in test with long string

To make the solver abort the length has to be strictly greater than
2^26.

romainbrenguier committed:
Deprecate string-max-length option

romainbrenguier committed:
Make char_array_of_pointer return a reference

romainbrenguier committed:
Move MAX_CONCRETE_STRING_SIZE definition to magic

Commit:
Extension to interpreter class

Commit:
Disable nested exception printing for Windows

Due to bug in the VS2013 C++ compiler, using std::rethrow_if_nested or
std::nested_exception is not supported. This disables trying to unwrap the
exception and just prints a warning saying the nested exceptionc couldn't be
printed.

Don't use noexcept directly, pull both part of the nested exception into
a separate file to handle discrepancies.

smowton committed:
Attribute main function arguments to __CPROVER_start

These are created to serve as arguments to main() (or whatever --function was passed), but are
declared in __CPROVER_start, and so should be named after their declarer like all other locals.

This was causing a problem in the security repository, where we relied on local variable names
being prefixed by their declaring function name, and the easiest fix was to restore that convention.

Commit:
Test demonstrating logging with clause for dealing with Windows

tautschnig committed:
Remove unused entries from the string table

21% of the entries do not occur in the code base (reduction from 844 to 663
entries).

tautschnig committed:
Use existing irep_idts instead of strings

tautschnig committed:
Check that the string table does not include unused entries

Commit:
Fix CMake build for Glucose Syrup

peterschrammel committed:
Bugfix: Java array symbols must have mode ID_java

peterschrammel committed:
Bugfix: Java symbol types must have mode ID_java

peterschrammel committed:
Replace asserts by invariants

peterschrammel committed:
Set mode in goto_convert auxiliary symbols

peterschrammel committed:
Bugfix: use proper language registration in unit tests

peterschrammel committed:
Utility functions to get mode and language

peterschrammel committed:
Associate dynamic objects with respective language mode

peterschrammel committed:
Add more doxygen to language.h and mode.h

Commit:
Add documentation to convert_bitvector

Commit:
Introduce exceptions for all conversion steps.

Commit:
Use format rather than from_expr for output

Commit:
Sort includes using clang-format

In current version of clang-format, this leaves groupings alone and just
sorts within the group.

For example, given a list of includes like:

```
include "boolbv.h"

include <util/std_types.h>
include <util/std_expr.h>
include <util/arith_tools.h>
include <util/base_type.h>
include <util/pointer_offset_size.h>

include <util/c_types.h>

```

These will be sorted to:

```
include "boolbv.h"

include <util/arith_tools.h>
include <util/base_type.h>
include <util/pointer_offset_size.h>
include <util/std_expr.h>
include <util/std_types.h>

include <util/c_types.h>
```

Note the "groups" are left untouched, the ordering occurs within the
groups

Commit:
Remove redundant default constructor

Commit:
Reformatting touched output function

tautschnig committed:
Make virtual function resolution independent of string table entry ordering

The program generated by virtual-function removal depended on where 'B', 'C',
'D', 'E' appear in the string table as there are multiple semantically
equivalent resolutions of virtual functions.

cesaro commented:
Is GH working? I cannot comment elsewhere!

cesaro commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -138,82 +254,266 @@ static void create_clinit_wrapper_symbols(
+/// static init has already taken place, calls the actual `<clinit>` method if
+/// not, and possibly recursively initializes super-classes and interfaces.
+/// Assume that C is the class to be initialized, then the algorithm is as
+/// follows:
```
Add `and that C extends C' and implements interfaces I1 ... In`.

cesaro commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -39,13 +67,35 @@ bool is_clinit_wrapper_function(const irep_idt &function_id)
+/// /param is_thread_local: if true this symbol will be set as thread local
+/// /param is_static_lifetime: if true this symbol will be set as static
+/// /return returns new symbol.
+static symbolt add_new_symbol(
```
I (partially) agree to the comment above, I think this method should call `get_fresh_aux_symbol` to insert the new symbol in the symbol table, but that method does not initialize the remaining fields as this method does (`is_lvalue`, `value` etc) in the manner required in this file, so I would leave this method here?

cesaro commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -14,6 +14,34 @@ Author: Chris Smowton, chris.smowton@diffblue.com
+/// `INIT_COMPLETE` corresponds to "the `<clinit>` has been executed and the
+///   class is ready to be used, or it has errored"
+///
+/// The last state corresponds to a fusion of the two original states "ready
```
@Degiorgio add the above sentence in the C++ comment.

cesaro commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -138,82 +254,266 @@ static void create_clinit_wrapper_symbols(
+///   // 'clinit_state' in order to ensure that only this thread runs the
+///   // static initializer. The assume() statement below will prevent the SAT
+///   // solver from producing a thread schedule where more than 1 thread is
+///   // initializing the initializer. At the end of this function the only
```
"running the initializer"

kroening committed:
fix array->f typechecking

Commit:
Update coding standard

Commit:
Remove brace checking from cpplint

Commit:
Remove NOLINTs for lambdas.

kroening committed:
allow address_of of byte_extract expressions

kroening committed:
show java_new_array_data side effects

tautschnig committed:
Remove unused symbolt::{to,from}_irep

Serialisation is done in write_goto_binary_v3 and read_bin_goto_object_v3.

tautschnig committed:
Move enum idt to the single translation unit that actually requires it

tautschnig committed:
Added an extension point for irep ids

tautschnig committed:
Fix Doxygen syntax

tautschnig committed:
array_size symbols: set mode and avoid redundant settings

The symbolt constructor already takes care of initialising fields to false.

tautschnig committed:
symex_dynamic::dynamic_object_size* are constants

We do not have a notion of guaranteed constants, thus mark them thread-local via
the auxiliary_symbolt constructor. This avoids repeated renaming of the size of
an array, which resulted in spurious errors in concurrent programs (e.g., in
SV-COMP running "cbmc --unwind 2 --stop-on-fail --32
c/pthread-C-DAC/pthread-numerical-integration_true-unreach-call.i").

Commit:
Rephrase and justify curly brace alignment exceptions

peterschrammel committed:
Use remove_java_new

peterschrammel committed:
Factor out java new removal into separate pass

peterschrammel committed:
Add Makefile dependency for smt2_solver

kroening committed:
more files to ignore

kroening committed:
use optional<size_t> instead of -1 in cmdlinet

tautschnig committed:
Use iosfwd instead of ostream where possible

tautschnig committed:
Use a single global INITIALIZE_FUNCTION macro instead of __CPROVER_initialize

tautschnig committed:
Remove unused includes

... and sort the remaining ones lexicographically.

tautschnig committed:
Use C++ streams instead of C-style snprintf

tautschnig committed:
Move asserts to invariants (and provide suitable includes)

tautschnig committed:
Do not unnecessarily use C string functions

tautschnig committed:
Use pointer_offset_bits instead of locally hacking up what it does anyway

tautschnig committed:
Move definition of base_type_eqt to .cpp

This is an implementation detail that need not be exposed to users - the
interface is the free function base_type_eq.

romainbrenguier committed:
Set mode of if_exprt introduced in preprocessing

A temporary variable is introduced later for this if_exprt and the mode
needs to be known for the variable, other an invariant may fail.

romainbrenguier committed:
Add test where tmp_if_expr is introduced

This checks the ID_mode fix for if_expr in preprocessing works correctly.

peterschrammel committed:
Assign mode to invalid objects

kroening committed:
remove support for SMT1

kroening committed:
remove OpenSMT

kroening committed:
whitespace fix

kroening committed:
remove do_unwind_module hook

kroening committed:
remove unused includes

kroening committed:
remove bmc_constraints

kroening committed:
missing const for parameter

kroening committed:
fix a comment

kroening committed:
cleanup the includes in src/java_bytecode

Degiorgio committed:
Added new cmd option to jbmc, 'java-threading'

Setting this option will enable  support for analysing
multi-threaded java programs. For now, this means that a thread-safe
version of clinit_wrapper is generated. This was made opt-in instead
of opt-out as the inclusion of thread-safe clinit wrapper has negative
implications for performance.

Commit also adds appropriate regression tests that use this flag to test
the resulting modifications to the clinit_wrapper
